### PR TITLE
Created OIDC module and examples

### PR DESCRIPTION
**Pull request by gautambaghel**

### CODE REVIEW

File: modules/tfc-oidc/variables.tf [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+variable "tfc_organization_name" {
+  type        = string
+  description = "The name of your Terraform Cloud organization"
+}
+
+variable "tfc_project_name" {
+  type        = string
+  default     = "Default Project"
+  description = "The project under which a workspace will be created"
```
Comment (danbarr) [inline modules/tfc-oidc/variables.tf]: Module isn't creating a workspace.

```suggestion
  description = "The Terraform Cloud project to authorize via OIDC"
```

File: modules/tfc-oidc/variables.tf [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+variable "tfc_workspace_name" {
+  type        = string
+  default     = "gcp-oidc-workspace"
+  description = "The name of the workspace that you'd like to create and connect to GCP"
```
Comment (danbarr) [inline modules/tfc-oidc/variables.tf]: Module isn't creating a workspace.

```suggestion
  description = "The name of the Terraform Cloud workspace to authorize via OIDC"
```

Additional suggestion...default this to `""` or note in the description that you can pass an empty string to authorize any workspace in the project (this supports self-service workspace creation using projects/varsets).

File: modules/tfc-oidc/main.tf [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+locals {
+  attribute_condition = var.attribute_condition == "" ? "assertion.sub.startsWith(\"organization:${var.tfc_organization_name}:project:${var.tfc_project_name}:workspace:${var.tfc_workspace_name}\")" : var.attribute_condition
+}
+
+# Enables the required services in the project.
+# https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/google_project_service
+resource "google_project_service" "services" {
+  count   = length(var.service_list)
+  service = var.service_list[count.index]
```
Comment (danbarr) [inline modules/tfc-oidc/main.tf]: Got this error here without adding the project_id: `Error: project: required field is not set`, also dependent service errors when destroying.

```suggestion
  count              = length(var.service_list)
  project            = var.project_id
  service            = var.service_list[count.index]
  disable_on_destroy = false
```

File: modules/tfc-oidc/main.tf [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+# Creates a workload identity pool to house a workload identity pool provider.
+# https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/iam_workload_identity_pool
+resource "google_iam_workload_identity_pool" "tfc_pool" {
+  provider                  = google-beta
```
Comment (danbarr) [inline modules/tfc-oidc/main.tf]: The beta provider isn't needed.

```suggestion
```

File: modules/tfc-oidc/main.tf [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  description               = var.pool_description
+  disabled                  = false
+}
+
+# Creates an identity pool provider which uses an attribute condition
+# to ensure that only the specified Terraform Cloud workspace will be
+# able to authenticate to GCP using this provider.
+# https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/iam_workload_identity_pool_provider
+resource "google_iam_workload_identity_pool_provider" "tfc_provider" {
+  provider                           = google-beta
```
Comment (danbarr) [inline modules/tfc-oidc/main.tf]: The beta provider isn't needed.

```suggestion
```

File: modules/tfc-oidc/versions.tf [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  required_providers {
+
+    google = {
+      source  = "hashicorp/google"
+      version = ">= 3.53, < 5.0.0"
+    }
+    google-beta = {
+      source  = "hashicorp/google-beta"
+      version = ">= 3.53, < 5.0.0"
+    }
```
Comment (danbarr) [inline modules/tfc-oidc/versions.tf]: The beta provider isn't needed.

```suggestion
```

File: examples/oidc-simple/versions.tf [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+terraform {
+  required_providers {
+    google = {
+      source  = "hashicorp/google"
+      version = "~> 4.0"
+    }
+    google-beta = {
+      source  = "hashicorp/google-beta"
+      version = "~> 4.0"
+    }
```
Comment (danbarr) [inline examples/oidc-simple/versions.tf]: The beta provider isn't needed.

```suggestion
```

**Review by danbarr — COMMENTED** [review]
Other items per our chat:

- Example should create the workspace variables for dynamic provider credentials
- Adding a random ID to the workload_identity_pool_id, at least in the example, will avoid errors due to GCP's 30-day hold on deleted pools

File: modules/tfc-oidc/main.tf [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  role               = "roles/iam.workloadIdentityUser"
+  member             = "principalSet://iam.googleapis.com/${google_iam_workload_identity_pool.tfc_pool.name}/${each.value.attribute}"
+}
+
+# Updates the IAM policy to grant the service account permissions within the project.
+# https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/google_project_iam
+resource "google_project_iam_member" "tfc_project_member" {
+  for_each = var.sa_mapping
+  project  = var.project_id
+  role     = "roles/editor"
```
Comment (danbarr) [inline modules/tfc-oidc/main.tf]: Should this be a variable instead of hard-coding the editor role?

Comment (gautambaghel) [reply modules/tfc-oidc/main.tf]: perhaps I should remove this entire role https://cloud.google.com/iam/docs/roles-overview

**Review by danbarr — APPROVED** [review]

**Pull request merged**